### PR TITLE
feat: support locale-specific customFormatters

### LINKED ISSUE #402: RFC: Support customFormatters locale specific modules
# Problem

The current custom formatter API does not allow me to import my runtime module with locale awareness.

Current: 
```js
export const phone: CustomFormatter = {
  // ...snip
  module: '@messageformat/phone'
};
```

Desired:
```
export const phone: CustomFormatter = {
  // ...snip
  module: (localeCode) => `@messageformat/upcase/${localeCode}` // e.g. "@messageformat/upcase/mx"
};
```

# Discussion

In this respect, the generated ESM javascript can use much smaller implementations for the compilation of interest!

### PR DESCRIPTION
**Pull request by cdaringe**
# Problem Statement

Given a locale & a message with a custom formatter requirement, I want my compiled output to use a formatter tailored to that locale, but cannot due to limitations in the current API.

Closes #402

## Motivating Context

Presume that some custom formatter is expensive to load for _all_ supported locales.

In the current API, my formatter must do something like this:

```js
formatter: (value, locale) => 
  isUs(locale)
  ? formatUs(value)
  : isMx(locale)
  ? formatMx(value)
//  ...SNIP N locales ...
  : isFinland(value)
  ? formatFn(value)
  : value
```

```js
// compiled-output.js
import { fmt } from "formatters/formatter-that-knows-about-all-locales-under-the-sun"
export const msg = d => fmt(d.foo)
```

 Presume that each formatting function is **expensive, say 1kb**. Now my app must install (N-1) kb of wasted formatters on each application load.

# Solution

Support locale-aware custom formatters.

```js
// compiled-output.js
import { fmt as fmt_enUS } from "formatters/en-US"
export const msg = d => fmt_enUS(d.foo)
```

```js
// compiled-output.js
import { fmt as fmt_esMX } from "formatters/en-MX"
export const msg = d => fmt_enMX(d.foo)
```

In the new state, customFormatters do not need to bundle _all_ locale concerns into a single module.

### CODE REVIEW

File: packages/core/package.json
Comment (cdaringe) [inline packages/core/package.json:17]: is that bad form 😄 ?

File: packages/core/src/compiler.ts
Comment (cdaringe) [inline packages/core/src/compiler.ts:207]: e.g. `es-US` => `esUS`

File: packages/core/src/compiler.ts
Comment (cdaringe) [inline packages/core/src/compiler.ts:323]: e.g. `setFormatter("phone__esUS", "phone")`

File: packages/core/src/messageformat.ts
Comment (cdaringe) [inline packages/core/src/messageformat.ts:84]: some of these keys could be deduped, but they were small enough i just dupe'd em (e.g. `arg`, `id`)

File: packages/core/src/compiler.ts
Comment (eemeli) [reply packages/core/src/compiler.ts:207]: Rather than putting together a new sanitization solution, it'd be better to use [`identifier()`](https://www.npmjs.com/package/safe-identifier#identifierkey-string-unique-boolean-string).

**Review by eemeli — COMMENTED** [review]
Took a quick glance, but I'll need to find time for a proper review. I'm a bit concerned about the expansion of state during the compilation; the `plural` arg is a bit badly named, but I would've thought that it was already tracking the current locale?

**Comment by cdaringe** [comment]
>  I'm a bit concerned about the expansion of state during the compilation

justifiably so! dig in at your convenience, let me know your thoughts

> the plural arg is a bit badly named, but I would've thought that it was already tracking the current locale?

the `plural/plurals` doesn't track locale. it has a `locale` field, but it's not honest. it's really just the lang, disguised as the locale.

for recollection:

```js
plurals = {
  // Record<LANG_STRING, PluralData>
en: { cardinals: ..., lc: 'en', locale: 'en' }
}
```

Problem is... i _need_ the actual/full locale code, e.g. `en-US`, not just the lang

**Comment by cdaringe** [comment]
Hey @eemeli, i know i'm blowing up your mentions this week 😄 . I'll go silent in all other forums if it helps! Would you be willing to give me some 👀 on this anytime soon?

**Comment by eemeli** [comment]
Hi, sorry for the delay here -- been on vacation (and still am...) & trying to stay afk. Back for a couple of days next week.

Reviewing this PR and how it fits in with the rest, I'm starting to get the sense that there's some premature caching happening with the `'*'` locale. This ends up being related to #361 as well. The current locale _should be_ available within the plural object. See for example the usage here: https://github.com/messageformat/messageformat/blob/04b395220c6f1327ede71fa5c888934ab8648e23/packages/core/src/compiler.ts#L185

Not sure if/when I'll have time for this myself, but for `'*'` the `plurals` array should probably be initialised not with all locales, but with the default locale and some marker indicating the wildcard. Then around here https://github.com/messageformat/messageformat/blob/04b395220c6f1327ede71fa5c888934ab8648e23/packages/core/src/compiler.ts#L81
the lookup should set the locale in the plural object according to the `key`.

**Comment by cdaringe** [comment]
No worries. Please, ignore this, and enjoy your vacation :)

> The current locale should be available within the plural object

Indeed! However, locale is not in actuality available. As I [mention in my comment above](https://github.com/messageformat/messageformat/pull/403#issuecomment-1591605774), `locale` is not actually `locale` at all. It's pretty much just `lang`, which is too lossy.

**Comment by eemeli** [comment]
`plural.locale` is only the primary subtag if you use `*` as the constructor locale; if you provide an array of locale identifiers there, they are used exactly.

This should probably be documented better.

**Comment by cdaringe** [comment]
Hm. Interesting. Shucks, I still feel like that my be a modeling issue in the existing plurals interface? I would ask "is a subtag a locale?". I'd suggest it is _not_. Either way, it's probably worth me learning OR us collaborating on what the correct nomenclature is for `lang + region` unambiguously. Perhaps `PluralObject` warrants some refinement?

> array should probably be initialised not with all locales, but with the default locale ... the lookup should set the locale in the plural object according to the key

i'm ...kinda following, but not 100%. i'm not confident enough to send a patch on this yet. i think you're talking about in the `Compiler::compile` method, maybe creating a new `PluralObject` on the fly with narrowed _stuff_.

If something were to happen with the above, would it be a mandatory precursor to this PR?

**Comment by eemeli** [comment]
> Hm. Interesting. Shucks, I feel like that there may be a modeling issue in the existing plurals interface? I would ask "is a subtag a locale?". I'd suggest it is _not_. Either way, it's probably worth me learning OR us collaborating on what the correct nomenclature is for `lang + region` unambiguously. Perhaps `PluralObject` warrants some refinement?

Not entirely sure what you're asking for here. To clarify, `plural.locale` is a BCP 47 language tag identifying a locale. Such a locale identifier requires at least a language tag, but may also include other subtags, such as a region identifier.

The plural objects are created during the MF constructor call, in two different ways:
1. If the given list of locales is explicit, we get an MF instance supporting exactly those locales.
2. If the user asks for all locales with `'*'`, plural objects are created for all supported _languages_, i.e. locales with only the primary tag (and pt-PT, but that's not relevant here).

The disparity in expectations arises when the second path is taken, _and_ a user uses `localeCodeFromKey` to identify a supported locale code. If that key is something that looks like a BCP 47 locale identifier (e.g. `en-CA` as a key mapping to the `en` locale code), they may expect for that _key_ to have some relevance for custom formatters, but it doesn't. Only the resolved locale code matters.

I don't think there's really a sensible way to get around this, and I'm not sure that it should be fixed. If you're doing complicated things, you should probably not use `'*'` but instead explicitly declare your supported locales.

> > array should probably be initialised not with all locales, but with the default locale ... the lookup should set the locale in the plural object according to the key
> 
> i'm ...kinda following, but not 100%. i'm not confident enough to send a patch on this yet. i think you're talking about in the `Compiler::compile` method, maybe creating a new `PluralObject` on the fly with narrowed _stuff_.
> 
> If something were to happen with the above, would it be a mandatory precursor to this PR?

No, that's optimisation that should be doable entirely separately. The next step with this PR is to get rid of the locale normalisation, additional state around that, and just use `plural.locale`. Also, please drop the docs updates; that gets pushed to the site from the main branch, so needs to wait for a version release.

**Review by eemeli — CHANGES_REQUESTED** [review]
See previous comment.

**Comment by cdaringe** [comment]
😵 Got it! Thanks for setting me straight. I now understand much better and see your perspective.

I've adapted the PR to your request, and it has shrunk up a bit 😄 !

File: packages/core/src/compiler.ts
Comment (eemeli) [inline packages/core/src/compiler.ts:175]: The option is always defined here.
```suggestion
        const formatter = this.options.customFormatters.[token.key];
```

File: packages/core/src/compiler.ts
Comment (eemeli) [inline packages/core/src/compiler.ts:179]: The temporary variable is more confusing than not.
```suggestion
        const isLocaleSpecificFormattingModule = typeof formatter?.module === 'function';
```

Also, "isLocaleSpecificFormattingModule" is an excessively long name for a local variable. With changes below, you can probably drop it rather than renaming it.

File: packages/core/src/compiler.ts
Comment (eemeli) [inline packages/core/src/compiler.ts:192]: This change needs to be reverted. We still want the locale as an argument because we should not assume that each locale will have an independent implementation. It should still be possible to e.g. say that all English (en, en-US, en-CA, en-UK, ...) use the same implementation, but possibly with some region-specific considerations.

File: packages/core/src/compiler.ts
Comment (eemeli) [reply packages/core/src/compiler.ts:315]: These arguments would be clearer for me with reversed order, i.e. "the `key` of the thing we're importing", and optionally "the `alias` we use for it internally". I'm also not a fan of really long names for local variables, esp. when combined with our 80 char width style.

**Review by eemeli — CHANGES_REQUESTED** [review]
Sorry for the delayed and partial review; too many things on my to-do list atm...

Only got part of the way through the compiler; haven't looked yet at what's happening with the `Object.assign`/`.bind()` stuff or anything further than that.

File: packages/core/src/compile-module.ts
Comment (cdaringe) [reply packages/core/src/compile-module.ts:34]: type narrowing only

File: packages/core/src/compiler.ts
Comment (cdaringe) [reply packages/core/src/compiler.ts]: 5bf4e84

File: packages/core/src/compiler.ts
Comment (cdaringe) [reply packages/core/src/compiler.ts]: 5bf4e84

**Comment by cdaringe** [comment]
1. locale is now always a param, per discussion
2. long var names compacted, per discussion

> [setFormatter] These arguments would be clearer for me with reversed order

This method is used in many places, and reversing the params would be a breaking change requiring a bulk edit in this module. Give me another review pass now that it's been through a first round of requested edits and assert that you still want this.

File: packages/core/src/compiler.ts
Comment (eemeli) [inline packages/core/src/compiler.ts:319]: Ok, I see now why mutating the formatter can be a bad idea: When it's not a built-in formatter that we're providing, it means we're modifying a value that a user has provided.

So using `.bind()` here is appropriate. See below for why you ought to inline the code for that, as you had it previously.

File: packages/core/src/compiler.ts
Comment (eemeli) [inline packages/core/src/compiler.ts:333]: OTOH, here the formatter is an internal one, so we don't need to clone it and can mutate it without ill effect.

File: packages/core/src/compiler.ts
Comment (eemeli) [inline packages/core/src/compiler.ts:325]: Checking `this.plural.locale` is incorrect here. If it may be null, then either it needs to be coerced to an empty string, or the function needs to be able to deal with a null input.

Fixing this will let you drop the `compile-module.ts` change.

File: packages/core/src/messageformat.ts
Comment (eemeli) [inline packages/core/src/messageformat.ts:57]: As we're now always including the locale code, am I right that the changes to this signature should be rolled back?

File: packages/core/src/messageformat.ts
Comment (eemeli) [inline packages/core/src/messageformat.ts:100]: Why does this expect the `locale` to be wrapped within an object? Wouldn't it be simpler to use `(locale: string) => string`?

**Review by eemeli — CHANGES_REQUESTED** [review]
Finally got through to the end. In general I think the approach is good, but we should be able to make the changes here more surgical; see inline comments.

**Comment by eemeli** [comment]
> > [setFormatter] These arguments would be clearer for me with reversed order
> 
> This method is used in many places, and reversing the params would be a breaking change requiring a bulk edit in this module. Give me another review pass now that it's been through a first round of requested edits and assert that you still want this.

Let's return to this after the current round of changes. My main concern is that with separate terms for the local variable name and the import name, those terms ought to correspond to the names used for the JS `import { foo as bar } ...` statement, where this is considered as "aliasing".

File: packages/core/src/messageformat.ts
Comment (cdaringe) [reply packages/core/src/messageformat.ts:100]: I had it that way originally, but i changed it to the object form so it could be extended without being a breaking change

File: packages/core/src/messageformat.ts
Comment (cdaringe) [reply packages/core/src/messageformat.ts:57]: No, otherwise I’m in no better position than when I started without the PR. the function form is precisely what lets me generate locale based module imports. The param is runtime, versus compile time

File: packages/core/src/messageformat.ts
Comment (eemeli) [reply packages/core/src/messageformat.ts:100]: Locale is good to keep as a separate argument. For example, that's the way it is in the MessageFormat constructor. Adding more arguments can be done later.

File: packages/core/src/messageformat.ts
Comment (eemeli) [reply packages/core/src/messageformat.ts:57]: The function form of `module` allows that. This is the signature of the function that'll be called during the formatting. After https://github.com/messageformat/messageformat/pull/403#discussion_r1272317743, it'll always include the locale argument.

File: packages/core/src/compile-module.ts
Comment (cdaringe) [reply packages/core/src/compile-module.ts]: removed! 72396da

File: packages/core/src/compiler.ts
Comment (cdaringe) [reply packages/core/src/compiler.ts]: ack. fixed, 72396da

File: packages/core/src/compiler.ts
Comment (cdaringe) [reply packages/core/src/compiler.ts]: ack, fixed 72396da

File: packages/core/src/messageformat.ts
Comment (cdaringe) [reply packages/core/src/messageformat.ts]: sure, i can get on board with that. refactored in ee31644

File: packages/core/src/messageformat.ts
Comment (cdaringe) [reply packages/core/src/messageformat.ts]: got it. i misinterpreted. my bad. addressed in a681a58

File: packages/core/src/compiler.ts
Comment (cdaringe) [reply packages/core/src/compiler.ts]: i wish we could do that, but TS doesn't allow you access an indexed field unless it's narrowed the type to an indexable type. `string` isn't indexable by string, so TS barfs. TS demands that you formatter to a form where `'module'` is guaranteed on an object form of `formatter`, so... no dice. nonetheless, in the spirit of your comment, i condensed it down to a single expression. a681a58

File: packages/core/src/compiler.ts
Comment (cdaringe) [reply packages/core/src/compiler.ts:334]: it's your call. i am biased to _not mutating the underlying fn_ ever, out of what I perceive to be general purpose best practices, but that is entirely subjective :)

**Comment by cdaringe** [comment]
Alrighty, pass #3, complete! Back to you!

File: packages/core/src/compiler.ts
Comment (eemeli) [inline packages/core/src/compiler.ts:180]:
```suggestion
        if (!formatter) {
```

File: packages/core/package.json
Comment (eemeli) [reply packages/core/package.json:17]: I don't really mind this, but please do at least match the format of the other entries.

File: packages/core/src/compiler.ts
Comment (eemeli) [reply packages/core/src/compiler.ts:334]: Let's just mutate it, and inline the `cloneFunction()` above.

**Review by eemeli — CHANGES_REQUESTED** [review]
Nearly there, I think. See inline replies.

File: packages/core/src/compiler.ts
Comment (cdaringe) [reply packages/core/src/compiler.ts]: b542f00

File: packages/core/src/compiler.ts
Comment (cdaringe) [reply packages/core/src/compiler.ts]: done, 9ee2bf1

**Comment by cdaringe** [comment]
pass 4 complete. back to you! thanks for the feedback. i think the PR is in much better shape that it was on first pass

**Review by eemeli — APPROVED** [review]
There are still a few code style issues in `setFormatter()`, but we can take care of them separately later.

The docs at `docs/custom-formatters.md` will need to get updated together with the next release. @cdaringe, would you be able/interested to submit a separate PR for that?

**Comment by cdaringe** [comment]
sound good. please see #408

**Pull request merged**
